README
NAME
    Filter::EOF - Run a callback after a file has been compiled

VERSION
    0.03

SYNOPSIS
      package MyPackage;
      use warnings;
      use strict;

      use Filter::EOF;

      sub import {
          my ($class, @args) = @_;
          my $caller = scalar caller;

          # set the COMPILE_TIME package var to a false value
          # when the file was compiled
          Filter::EOF->on_eof_call(sub {
              no strict 'refs';
              ${ $caller . '::COMPILE_TIME' } = 0;
          });

          # set the COMPILE_TIME package var to a true value when
          # we start compiling it.
          {   no strict 'refs';
              ${ $caller . '::COMPILE_TIME' } = 1;
          }
      }

      1;
      ...

      package MyUsingPackage;
      use warnings;
      use strict;

      our $COMPILE_TIME;
      use MyPackage;

      # prints 'yes'
      BEGIN { print +( $COMPILE_TIME ? 'yes' : 'no' ), "\n" }

      # prints 'no'
      print +( $COMPILE_TIME ? 'yes' : 'no' ), "\n";

      1;

DESCRIPTION
    This module utilises Perl's source filters to provide you with a
    mechanism to run some code after a file using your module has been
    processed.

METHODS
  "import( @functions )"
    Currently, only a function equivalent of the "on_eof_call" method is
    provided for export.

      use Filter::EOF qw( on_eof_call );

      sub import {
          my ($class) = @_;
          ...
          on_eof_call { ... };
      }
      ...

  "on_eof_call( $code_ref )"
    Call this method in your own "import" method to register a code
    reference that should be called when the file "use"ing yours was
    compiled.

    The code reference will get a scalar reference as first argument to an
    empty string. if you change this string to something else, it will be
    appended at the end of the source.

      # call C<some_function()> after runtime.
      Filter->on_eof_call(sub { 
          my $append = shift;
          $$append .= '; some_function(); 1;';
      });

EXPORTS
  on_eof_call
    You can optionally import the "on_eof_call" function into your
    namespace.

EXAMPLES
    You can find the example mentioned in "SYNOPSIS" in the distribution
    directory "examples/synopsis/".

SEE ALSO
    Filter::Call::Util, "Exporting without using Exporter's import method"
    in Exporter

AUTHOR AND COPYRIGHT
    Robert 'phaylon' Sedlacek - "<rs@474.at>". Many thanks to Matt S Trout
    for the idea and inspirations on this module.

LICENSE
    This program is free software; you can redistribute it and/or modify it
    under the same terms as perl itself.



Changes

0.03 / Sun Jun 24 18:35:07 CEST 2007
    - Added on_eof_call export

0.02 / Sat Feb  3 21:46:45 CET 2007
    - Some POD corrections
    - Appending code after eof implemented

0.01 / Sun Jan 21 19:28:49 CET 2007
    - Initial version


t/lib/SimpleTest.pm
package SimpleTest;
use warnings;
use strict;

our (
    $TEST_ON_IMPORT,
    $TEST_ON_IMPORT_CLASS,
    $TEST_ON_IMPORT_ARGS,

    $TEST_ON_EOF,
    $TEST_ON_EOF_CLASS,

    $TEST_ORDER,
    $TEST_PHASE,
    $TEST_COMPILETIME,
    $TEST_RUNTIME,
);

use SimpleFilter foo => 23, bar => 13;

BEGIN { $TEST_COMPILETIME = 1 if $TEST_PHASE eq 'compile' }

$TEST_RUNTIME = 1 if $TEST_PHASE eq 'run';

1;


t/lib/SimpleFilter.pm
package SimpleFilter;
use warnings;
use strict;

use Filter::EOF;
use base 'Exporter';

our @EXPORT = qw(test_export);

sub import {
    my ($class, @args) = @_;
    Filter::EOF->on_eof_call(sub { 

        $SimpleTest::TEST_ON_EOF       = 1;
        $SimpleTest::TEST_ON_EOF_CLASS = $class;
        $SimpleTest::TEST_PHASE = 'run';

        push @{ $SimpleTest::TEST_ORDER }, 'on_eof';
    });

    $SimpleTest::TEST_ON_IMPORT       = 1;
    $SimpleTest::TEST_ON_IMPORT_CLASS = $class;
    $SimpleTest::TEST_ON_IMPORT_ARGS  = \@args;
    $SimpleTest::TEST_PHASE = 'compile';

    $SimpleTest::TEST_ORDER = ['on_import'];

    $class->export_to_level(1);
}

sub test_export { 23 }

1;


lib/Filter/EOF.pm
package Filter::EOF;

=head1 NAME

Filter::EOF - Run a callback after a file has been compiled

=cut

use warnings;
use strict;

=head1 VERSION

0.03

=cut

our $VERSION = 0.03;

=head1 SYNOPSIS

  package MyPackage;
  use warnings;
  use strict;

  use Filter::EOF;

  sub import {
      my ($class, @args) = @_;
      my $caller = scalar caller;

      # set the COMPILE_TIME package var to a false value
      # when the file was compiled
      Filter::EOF->on_eof_call(sub {
          no strict 'refs';
          ${ $caller . '::COMPILE_TIME' } = 0;
      });

      # set the COMPILE_TIME package var to a true value when
      # we start compiling it.
      {   no strict 'refs';
          ${ $caller . '::COMPILE_TIME' } = 1;
      }
  }

  1;
  ...

  package MyUsingPackage;
  use warnings;
  use strict;

  our $COMPILE_TIME;
  use MyPackage;

  # prints 'yes'
  BEGIN { print +( $COMPILE_TIME ? 'yes' : 'no' ), "\n" }

  # prints 'no'
  print +( $COMPILE_TIME ? 'yes' : 'no' ), "\n";

  1;

=head1 DESCRIPTION

This module utilises Perl's source filters to provide you with a
mechanism to run some code after a file using your module has been
processed.

=cut

use Carp;
use Filter::Util::Call;

=head1 METHODS

=head2 C<import( @functions )>

Currently, only a function equivalent of the C<on_eof_call> method
is provided for export.

  use Filter::EOF qw( on_eof_call );

  sub import {
      my ($class) = @_;
      ...
      on_eof_call { ... };
  }
  ...

=cut

my %Export = (
    on_eof_call => sub { my $class = shift; sub (&) { $class->on_eof_call(@_) } },
);

sub import {
    my ($class, @names) = @_;
    for my $name (@names) {
        Carp::croak "Unknown function '$name'" unless exists $Export{ $name };
        no strict 'refs';
        *{ scalar(caller) . '::' . $name } = $Export{ $name }->($class);
    }
    return 1;
}

=head2 C<on_eof_call( $code_ref )>

Call this method in your own C<import> method to register a code 
reference that should be called when the file C<use>ing yours was
compiled.

The code reference will get a scalar reference as first argument
to an empty string. if you change this string to something else,
it will be appended at the end of the source.

  # call C<some_function()> after runtime.
  Filter->on_eof_call(sub { 
      my $append = shift;
      $$append .= '; some_function(); 1;';
  });

=cut

sub on_eof_call {
    my ($class, $code) = @_;

    my $past_eof;
    filter_add( sub {
        my $status = filter_read;
        return $status if $past_eof;
        if ($status == 0) {
            $code->(\$_);
            $status   = 1;
            $past_eof = 1;
        }
        return $status;
    });
}

=head1 EXPORTS

=head2 on_eof_call

You can optionally import the C<on_eof_call> function into your namespace.

=head1 EXAMPLES

You can find the example mentioned in L</SYNOPSIS> in the distribution
directory C<examples/synopsis/>.

=head1 SEE ALSO

L<Filter::Call::Util>,
L<Exporter/Exporting without using Exporter's import method>

=head1 AUTHOR AND COPYRIGHT

Robert 'phaylon' Sedlacek - C<E<lt>rs@474.atE<gt>>. Many thanks to
Matt S Trout for the idea and inspirations on this module.

=head1 LICENSE

This program is free software; you can redistribute it and/or modify
it under the same terms as perl itself.

=cut

1;
